### PR TITLE
Reduce global lock use for ConcurrentDictionary

### PR DESCRIPTION
**Pull request by pentp**
* Reduce global locking when accessing `ConcurrentDictionary` by reducing use of `Keys/Values/Count/IsEmpty`.
* Remove some dead code.

Inspider by #6870.

### CODE REVIEW

File: src/Orleans.Runtime/Networking/ConnectionListener.cs [hunk 1]
```diff
@@ -131,23 +128,23 @@ public async Task StopAsync(CancellationToken cancellationToken)
                 }
 
                 var cycles = 0;
-                var closeTasks = new List<Task>();
                 var cancellationTask = cancellationToken.WhenCancelled();
-                while (this.ConnectionCount > 0)
+                for (var closeTasks = new List<Task>(); ; closeTasks.Clear())
```
Comment (ReubenBond) [inline src/Orleans.Runtime/Networking/ConnectionListener.cs:132]: I think this construct is not as clear as the boring alternative

File: src/Orleans.Streaming/Internal/StreamConsumerExtension.cs [hunk 2]
```diff
@@ -187,17 +187,17 @@ public Task<StreamHandshakeToken> GetSequenceToken(GuidId subscriptionId)
...
+        internal List<StreamSubscriptionHandleImpl<T>> GetAllStreamHandles<T>()
         {
-            return allStreamObservers.Values
-                .OfType<StreamSubscriptionHandleImpl<T>>()
-                .Where(o => o != null)
-                .ToList();
+            var ls = new List<StreamSubscriptionHandleImpl<T>>();
+            foreach (var o in allStreamObservers)
+            {
+                if (o.Value is StreamSubscriptionHandleImpl<T> i) ls.Add(i);
```
Comment (ReubenBond) [inline src/Orleans.Streaming/Internal/StreamConsumerExtension.cs:198]: We should use the `is {} i` pattern to make ourselves immune to refactoring. Eg, if allStreamObservers' element type changes, then this condition may always be false.

File: src/Orleans.Streaming/SimpleMessageStream/SimpleMessageStreamProducerExtension.cs [hunk 3]
```diff
@@ -184,10 +183,10 @@ internal void AddRemoteSubscriber(GuidId subscriptionId, IStreamConsumerExtensio
...
             {
                 consumers.TryRemove(subscriptionId, out _);
-                if (consumers.Count == 0)
-                {
-                    // Unsubscribe from PubSub?
-                }
+                //if (consumers.IsEmpty)
+                //{
+                //    // Unsubscribe from PubSub?
+                //}
```
Comment (ReubenBond) [inline src/Orleans.Streaming/SimpleMessageStream/SimpleMessageStreamProducerExtension.cs:189]: Prefer we delete this code instead, potentially leaving the original comment the note for future spelunkers

**Review by ReubenBond — APPROVED** [review]

File: src/Orleans.Streaming/Internal/StreamConsumerExtension.cs (hunk 2 above)
Comment (ReubenBond) [reply src/Orleans.Streaming/Internal/StreamConsumerExtension.cs:198]: My mistake - this is a downcast

File: src/Orleans.Runtime/Networking/ConnectionListener.cs (hunk 1 above)
Comment (pentp) [reply src/Orleans.Runtime/Networking/ConnectionListener.cs:132]: Fair enough. Both look a bit awkward to me, but I think I used this version here as I introduced a similar loop into ConnectionManager (where previously each iteration created a new list).

I agree that it's an uncommon for-loop pattern, but it avoids both the possibility of accidentally skipping the `Clear` call and avoids the call for the first iteration (which in this case really doesn't matter).

**Pull request merged**